blockchain/chain.go

```diff
@@ -1665,6 +1665,82 @@ func (b *BlockChain) locateHeaders(locator BlockLocator, hashStop *chainhash.Has
 	return headers
 }
 
+// ReconsiderBlock takes a block hash and allows it to be revalidated.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) ReconsiderBlock(hash *chainhash.Hash) error {
+	b.chainLock.Lock()
+	defer b.chainLock.Unlock()
+	return b.reconsiderBlock(hash)
+}
+
+// reconsiderBlock takes a block hash and allows it to be revalidated.
+func (b *BlockChain) reconsiderBlock(hash *chainhash.Hash) error {
+	node := b.index.LookupNode(hash)
+	if node == nil {
+		err := fmt.Errorf("block %s is not known", hash)
+		return err
+	}
+
+	// No need to reconsider, it is already valid
+	if node.status.KnownValid() {
+		return nil
+	}
+
+	var blk *bchutil.Block
+
+	// Find fork point if one exists
+	forkNode := b.bestChain.FindFork(node)
+	if forkNode != nil {
+		// Roll back to the fork point and load all the nodes into a list
+		nodes := list.New()
+		for n := node; n != nil && n != forkNode; n = n.parent {
+			nodes.PushFront(n)
+		}
+
+		// Iterate over the list in reverse order so that the forkNode is processed first
+		for e := nodes.Front(); e != nil; e = e.Next() {
```

Comment:
Still think this needs to be iterated over this way:
```
var next *list.Element
	for e := nodes.Front(); e != nil; e = next {
		next = e.Next()
```
You'll note in the example if you just use e = e.Next() then e is nil on the second iteration through the loop and the loop exists.